pkg/distributor/validate.go

```diff
@@ -149,6 +150,50 @@ func (e labelValueTooLongError) Error() string {
 	return fmt.Sprintf(labelValueTooLongMsgFormat, len(e.Label.Value), e.Limit, e.Label.Name, e.Label.Value, e.Series)
 }
 
+// labelValueTooLongSummaries holds a summary for each metric and label processed
+// in a write request that have values exceeding the limit.
+type labelValueTooLongSummaries struct {
+	globalCount int
+	summaries   []labelValueTooLongSummary
+}
+
+func (s *labelValueTooLongSummaries) measure(metric, label, originalValue mimirpb.UnsafeMutableString) {
+	if s == nil {
+		return
+	}
+	s.globalCount++
+	if s.summaries == nil {
+		s.summaries = make([]labelValueTooLongSummary, 0, labelValueTooLongSummariesLimit)
+	}
+
+	i := slices.IndexFunc(s.summaries, func(summary labelValueTooLongSummary) bool {
+		return summary.metric == metric && summary.label == label
+	})
+	if i != -1 {
+		s.summaries[i].count++
+		return
+	}
+
+	if len(s.summaries) >= labelValueTooLongSummariesLimit {
+		return
+	}
+
+	s.summaries = append(s.summaries, labelValueTooLongSummary{
+		metric:            metric,
+		label:             label,
+		sampleValue:       fmt.Sprintf("%.200s (truncated)", originalValue),
+		sampleValueLength: len(originalValue),
+		count:             1,
+	})
+}
+
+type labelValueTooLongSummary struct {
+	metric, label     mimirpb.UnsafeMutableString
```

Comment:
I realized that these are unsafe, let's just clone them into safe strings?
```suggestion
	metric, label     string
```